bemani/backend/iidx/cannonballers.py

```diff
@@ -1,15 +1,2184 @@
+                data.set_attribute('score', str(score[1].points))
+                data.set_attribute('clflg', str(self.db_to_game_status(score[1].data.get_int('clear_status'))))
+                data.set_attribute('pid', str(profile.get_int('pid')))
+                data.set_attribute('myFlg', '1' if score[0] == userid else '0')
+                data.set_attribute('update', '0')
+
+                data.set_attribute('sgrade', str(
+                    self.db_to_game_rank(profile.get_int(self.DAN_RANKING_SINGLE, -1), self.GAME_CLTYPE_SINGLE),
+                ))
+                data.set_attribute('dgrade', str(
+                    self.db_to_game_rank(profile.get_int(self.DAN_RANKING_DOUBLE, -1), self.GAME_CLTYPE_DOUBLE),
+                ))
+
+                qpro = profile.get_dict('qpro')
+                data.set_attribute('head', str(qpro.get_int('head')))
+                data.set_attribute('hair', str(qpro.get_int('hair')))
+                data.set_attribute('face', str(qpro.get_int('face')))
+                data.set_attribute('body', str(qpro.get_int('body')))
+                data.set_attribute('hand', str(qpro.get_int('hand')))
+
+                record_num = record_num + 1
+
+        return root
+
+    def handle_IIDX25music_play_request(self, request: Node) -> Node:
+        musicid = int(request.attribute('mid'))
+        chart = int(request.attribute('clid'))
+        clear_status = self.game_to_db_status(int(request.attribute('cflg')))
+
+        self.update_score(
+            None,  # No userid since its anonymous
+            musicid,
+            chart,
+            clear_status,
+            0,  # No ex score
+            0,  # No ex score
+            0,  # No miss count
+            None,  # No ghost
+            None,  # No shop for this user
+        )
+
+        # Calculate and return statistics about this song
+        root = Node.void('IIDX25music')
+        root.set_attribute('clid', request.attribute('clid'))
+        root.set_attribute('mid', request.attribute('mid'))
+
+        attempts = self.get_clear_rates(musicid, chart)
+        count = attempts[musicid][chart]['total']
+        clear = attempts[musicid][chart]['clears']
+        full_combo = attempts[musicid][chart]['fcs']
+
+        if count > 0:
+            root.set_attribute('crate', str(int((1000 * clear) / count)))
+            root.set_attribute('frate', str(int((1000 * full_combo) / count)))
+        else:
+            root.set_attribute('crate', '0')
+            root.set_attribute('frate', '0')
+
+        return root
+
+    # Bare minimum response to say we handled the request
+    def handle_IIDX25music_beginnerplay_request(self, request: Node) -> Node:
+        return Node.void('IIDX25music')
+
+    def handle_IIDX25music_arenaCPU_request(self, request: Node) -> Node:
+        root = Node.void('IIDX25music')
+        music_list = [0, 0, 0, 0]
+        cpu_list = [0, 0, 0, 0]
+        grade_to_percentage_map = {
```

Comment:
Where is this from?

Reply:
I pretty much just made up these values on the spot . Would it be better to make the constants on the left use the DAN_RANK constants? All this method exists to do is allow the CPUs to do something rather than just sit still

Reply:
Are they the DAN rank IDs on the left side? If so, then yes we should use those constants for documentation and such.